test/hotspot/jtreg/compiler/jvmti/TriggerBuiltinExceptionsTest.java

```diff
@@ -0,0 +1,132 @@
+        try {
+            int idxFromSrc = (int)(src[i]);                   // NPEs, CastExceptions;
+            int rangeAdjust = 3 / (idxFromSrc % 3);           // Each 3rd is division by 0
+            Object value = src[i - rangeAdjust];              // Array indexing is broken
+            dest[i] = (Long)value;                            // ArrayStoreException or CastException
+        } catch (Exception e) {
+            caughtByJavaTest += 1;
+        }
+    }
+
+    /**
+     * Makes sure that method is compiled.
+     */
+    private static void compileMethodOrThrow(Method method) {
+        boolean enqueued = WB.enqueueMethodForCompilation(method, COMP_LEVEL);
+        if (!enqueued) {
+            throw new Error(String.format("%s can't be enqueued for compilation on level %d",
+                        method, COMP_LEVEL));
+        }
+        Asserts.assertTrue(
+                Utils.waitForCondition(() -> WB.isMethodCompiled(method), COMP_TIMEOUT),
+                String.format("Method hasn't been compiled in %d millis", COMP_TIMEOUT));
+    }
+
+    /**
+     * 1. Compiles method with no profiling information;
+     * 2. Causes deoptimization using arguments that causes hot throws;
+     * 3. Compiles the method again;
+     * 4. Checks that exceptions within compiled code are registered by JVMTI agent.
+     */
+    public static void main(String[] args) throws Throwable {
+        // Preparing the method
+        final Method method = TriggerBuiltinExceptionsTest.class.getMethod(
+                "methodToCompile", int.class, Object[].class, Object[].class);
+        WB.deoptimizeMethod(method);
+        TriggerBuiltinExceptionsTest.compileMethodOrThrow(method);
+
+        // Preparing source - badly formed, supposed-to-be-int array
+        final Object[] src = new Object[] {
+                valueOf(0), null, null, valueOf(3), null, valueOf(5), "string",
+                Long.valueOf(Long.MAX_VALUE), null, valueOf(9), "string"};
+        final Object[] dst = new Integer[ITERATIONS];
+
+        // 1. Should cause deoptimization as array is null
+        // 2. Make the throw hot to make the next compilation aware of it.
+        for (int i = 0; i < ITERATIONS; i++) {
+            TriggerBuiltinExceptionsTest.methodToCompile(i, src, dst);
+        }
+        Asserts.assertTrue(
+                Utils.waitForCondition(() -> !WB.isMethodCompiled(method), COMP_TIMEOUT),
+                String.format("Method hasn't been deoptimized in %d millis", COMP_TIMEOUT));
+
+        // Compile method with throw being hot
+        TriggerBuiltinExceptionsTest.compileMethodOrThrow(method);
+
+        // Gathering exceptions in compiled code
+        for (int i = 0; i < ITERATIONS; i++) {
+            TriggerBuiltinExceptionsTest.methodToCompile(i, src, dst);
+        }
+
+        Asserts.assertEQ(
+                TriggerBuiltinExceptionsTest.caughtByJVMTIAgent(), caughtByJavaTest,
```

Comment:
What is the reason to use the class name prefix for methods? :
  TriggerBuiltinExceptionsTest.compileMethodOrThrow
  TriggerBuiltinExceptionsTest.methodToCompile
  TriggerBuiltinExceptionsTest.caughtByJVMTIAgent
It is not really needed, tight?

Reply:
Style habits, acquired in previons job... fixed.